Lib/idlelib/editor.py

```diff
@@ -1368,14 +1365,18 @@ def smart_backspace_event(self, event):
         have = len(chars.expandtabs(tabwidth))
         assert have > 0
         want = ((have - 1) // self.indentwidth) * self.indentwidth
-        # Debug prompt is multilined....
+
         ncharsdeleted = 0
-        while True:
-            chars = chars[:-1]
-            ncharsdeleted = ncharsdeleted + 1
-            have = len(chars.expandtabs(tabwidth))
-            if have <= want or chars[-1] not in " \t":
+        have = len(chars.expandtabs(tabwidth))
+
+        for i in range(len(chars) - 1, -1, -1):
+            if have <= want or chars[i] not in " \t":
```

Comment:
`have <= want` is now a constant condition but `have` previoulsy was recomputed at every iteration.

Reply:
Well, that's the reason we should add tests. Note that the logic is as follows:

- we want to remove trailing whitespaces but we don't know how many we want to remove
- we *only* remove trailing whitespaces; if we encounter something else, we stop
- every time we remove a trailing whitespace, we recompute `len(chars.expandtabs(tabwidth))`. If we compute this before the loop and store it in `have`, we can simply remove tabwidth from `have` if we encounter a `\t`.